File: LunchApp/app/src/main/java/br/com/gersonsilvafilho/lunchapp/data/InMemoryRestaurantRepository.java

package br.com.gersonsilvafilho.lunchapp.data;

import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;

import com.google.common.collect.ImmutableList;

import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Concrete implementation to load Restaurants from the a data source.
 */
public class InMemoryRestaurantRepository implements RestaurantRepository {

    private final RestaurantServiceApi mRestaurantServiceApi;

    /**
     * This method has reduced visibility for testing and is only visible to tests in the same
     * package.
     */
    @VisibleForTesting
    List<Restaurant> mCachedRestaurants;

    public InMemoryRestaurantRepository(@NonNull RestaurantServiceApi restaurantServiceApi) {
        mRestaurantServiceApi = checkNotNull(restaurantServiceApi);
    }

    @Override
    public void getRestaurants(@NonNull final LoadRestaurantsCallback callback) {
        checkNotNull(callback);
        // Load from API only if needed.
        if (mCachedRestaurants == null) {
            mRestaurantServiceApi.getAllRestaurant(new RestaurantServiceApi.RestaurantsServiceCallback<List<Restaurant>>()
            {
                @Override
                public void onLoaded(List<Restaurant> restaurants) {
                    mCachedRestaurants = ImmutableList.copyOf(restaurants);
                    callback.onRestaurantsLoaded(mCachedRestaurants);
                }
            });
        } else {
            callback.onRestaurantsLoaded(mCachedRestaurants);
        }
    }

    @Override
    public void saveRestaurant(@NonNull Restaurant restaurant) {
        checkNotNull(restaurant);
        mRestaurantServiceApi.saveRestaurant(restaurant);
        refreshData();
    }

    @Override
    public void getRestaurant(@NonNull final String RestaurantId, @NonNull final GetRestaurantCallback callback) {
        checkNotNull(RestaurantId);
        checkNotNull(callback);
        // Load Restaurants matching the id always directly from the API.
        mRestaurantServiceApi.getRestaurant(RestaurantId, new RestaurantServiceApi.RestaurantsServiceCallback<Restaurant>()
        {
            @Override
            public void onLoaded(Restaurant restaurant) {
                callback.onRestaurantsLoaded(restaurant);
            }
        });
    }

    @Override
    public void refreshData() {
        mCachedRestaurants = null;
    }

}

File: LunchApp/app/src/main/java/br/com/gersonsilvafilho/lunchapp/data/InMemoryRestaurantRepository.kt
package br.com.gersonsilvafilho.lunchapp.data

import android.graphics.Bitmap
import android.support.annotation.VisibleForTesting
import com.google.common.base.Preconditions.checkNotNull
import com.google.common.collect.ImmutableList
import java.util.*

/**
 * Concrete implementation to load Restaurants from the a data source.
 */
class InMemoryRestaurantRepository(restaurantServiceApi: RestaurantServiceApi) : RestaurantRepository {

    private val mRestaurantServiceApi: RestaurantServiceApi

    /**
     * This method has reduced visibility for testing and is only visible to tests in the same
     * package.
     */
    @VisibleForTesting
    var mCachedRestaurants: List<Restaurant>? = null

    init {
        mRestaurantServiceApi = checkNotNull(restaurantServiceApi)
    }

    override fun getRestaurants(date: Date, callback: RestaurantRepository.LoadRestaurantsCallback) {
        checkNotNull(callback)
        // Load from API only if needed.
        if (mCachedRestaurants == null) {
            mRestaurantServiceApi.getAllRestaurant(date, object : RestaurantServiceApi.RestaurantsServiceCallback<List<Restaurant>> {
                override fun onLoaded(restaurants: List<Restaurant>) {
                    mCachedRestaurants = ImmutableList.copyOf(restaurants)
                    callback.onRestaurantsLoaded(mCachedRestaurants!!)
                }
            })
        } else {
            callback.onRestaurantsLoaded(mCachedRestaurants!!)
        }
    }

    override fun saveRestaurant(restaurant: Restaurant) {
        checkNotNull(restaurant)
        mRestaurantServiceApi.saveRestaurant(restaurant)
        refreshData()
    }

    override fun getRestaurant(RestaurantId: String, callback: RestaurantRepository.GetRestaurantCallback) {
        checkNotNull(RestaurantId)
        checkNotNull(callback)
        // Load Restaurants matching the id always directly from the API.
        mRestaurantServiceApi.getRestaurant(RestaurantId, object : RestaurantServiceApi.RestaurantsServiceCallback<Restaurant> {
            override fun onLoaded(restaurant: Restaurant) {
                callback.onRestaurantsLoaded(restaurant)
            }
        })
    }

    override fun refreshData() {
        mCachedRestaurants = null
    }

    override fun getRestaurantImageBitmap(RestaurantId: String, imgHeight: Int, imageWidth: Int, callback: RestaurantRepository.GetRestaurantImageCallback) {
        checkNotNull(String)
        checkNotNull(callback)
        // Load Restaurants matching the id always directly from the API.
        mRestaurantServiceApi.getRestaurantImageBitmap(RestaurantId, imgHeight, imageWidth, object : RestaurantServiceApi.RestaurantsServiceCallback<Bitmap>
        {
            override fun onLoaded(bitmap: Bitmap) {
                callback.onRestaurantImageLoaded(bitmap)
            }
        })
    }

    override fun sendRestaurantVote(RestaurantId: String, UserId: String, date: Date, callback: RestaurantRepository.SendRestaurantVoteCallback)
    {
        checkNotNull(String)
        checkNotNull(callback)

        mRestaurantServiceApi.sendRestaurantVote(RestaurantId, UserId, date, object : RestaurantServiceApi.RestaurantsServiceCallback<Map<String,String>>
        {
            override fun onLoaded(voteIsOk: Map<String,String>)
            {
                callback.onRestaurantVote(voteIsOk)
            }
        })
    }

}
